### ISSUE TITLE
Blazor Circuit has been shut down due to error after upgrading to 3.1.8 in Dispose When _JSRuntime.InvokeVoidAsync

### ISSUE THREAD

**Issue by lucianparvu** [dotnet/aspnetcore#25907]
### Describe the bug
After upgrading to Net Core: 3.1.8 (Blazor Server Side)  when invoke  some javascript functions in Dispose Method and I refresh the browser I get : Blazor Circuit has been shut down due to error . This is frustrating because in the same time the div components-reconnect-modal is displayed an then disappear.

I have tested into another server with 3.1.7 and there is no error .
 
```
 public void Dispose()
    {    
        try
        {
            if (!string.IsNullOrEmpty(Src))
            {
 		Task.Run(() => _JsRuntime.InvokeVoidAsync("UnLoadJavaScriptByUrl", Src)).GetAwaiter().GetResult();                     
            }
        }
        catch /*(Exception ex)*/
        {

        }
    }
```

**How to Reproduce:**
- In a Default Blazor Server Template in index.razor
```
@page "/"
@implements IDisposable
@inject IJSRuntime _JsRuntime
 
@code {

    public void Dispose()
    {
        Task.Run(() => _JsRuntime.InvokeVoidAsync("UnLoadJavaScriptByUrl")).GetAwaiter().GetResult();
    }
}
 
```

In _Host.cshtml, after  <script src="_framework/blazor.server.js"></script> add:

```
 <script>
        window.UnLoadJavaScriptByUrl = function () {
            console.log('test');
        };
    </script>
```

If you set a debugger point on Task.Run(() => _JsRuntime.InvokeVoidAsync("UnLoadJavaScriptByUrl")).GetAwaiter().GetResult(); and refresh the page after the line executes you will see the next error in console . In the Hello World application it is happens too fast, but in a real application you can see the error for 1-2 seconds .

Please see the print screen

![Capture](https://user-images.githubusercontent.com/26578147/93240579-f5d3c500-f78c-11ea-99bd-0093a4b15ef0.JPG)

**Comment by mkArtakMSFT** [comment]
Thank you for filing this issue. In order for us to investigate this issue, please provide a minimalistic repro project (ideally a GitHub repo) that illustrates the problem.

**Comment by lucianparvu** [comment]
> Thank you for filing this issue. In order for us to investigate this issue, please provide a minimalistic repro project (ideally a GitHub repo) that illustrates the problem.

I have updated the First Comment Description .

**Comment by mkArtakMSFT** [comment]
Can you try to avoid a blocking call by changing your code to:

```csharp
public async void Dispose()
{
      try
        {
            if (!string.IsNullOrEmpty(Src))
            {
 		await _JsRuntime.InvokeVoidAsync("UnLoadJavaScriptByUrl", Src);
            }
        }
        catch /*(Exception ex)*/
        {

        }
}
```

**Comment by lucianparvu** [comment]
Thank's .

I have tried, but on full page refresh (From Browser) the error and the div blazor-error-ui is still displayed .

**Comment by handirusli** [comment]
Is there a way to rollback to version 3.1.7 ?
I have a similar problem when invoke javascript function on OnAfterRender()

**Comment by ADefWebserver** [comment]
> Is there a way to rollback to version 3.1.7 ?

I ask the same thing here: https://github.com/dotnet/aspnetcore/issues/25646#issuecomment-704648884

**Comment by SteveSandersonMS** [comment]
I think this problem will go away when the `beforeunload` handler fix is rolled out. Do you agree, @captainsafia / @pranavkm? Should we be able to close this?

**Comment by captainsafia** [comment]
@SteveSandersonMS It appears so based on the behavior described here (disconnect on page refresh).

@lucianparvu This issue should be resolved in .NET 5 RC2, which ships on October 13th. If you have the chance, please try that build to resolve your issue.

You can also try installing the nightly build of RC2 from https://github.com/dotnet/installer#installers-and-binaries and test that out.

We'll be backporting a fix for this to 3.1 in the next patch release.

**Comment by lucianparvu** [comment]
> @SteveSandersonMS It appears so based on the behavior described here (disconnect on page refresh).
> 
> @lucianparvu This issue should be resolved in .NET 5 RC2, which ships on October 13th. If you have the chance, please try that build to resolve your issue.
> 
> You can also try installing the nightly build of RC2 from https://github.com/dotnet/installer#installers-and-binaries and test that out.
> 
> We'll be backporting a fix for this to 3.1 in the next patch release.

Thank's

We have tried  to upgrade 2 days ago to NET 5 RC1, but we have a lot of issues and errors because of EF Core Left Joins . It seems in NET 5  you can't do some queries like :

```
context.Products.Select(s => new Product {
                    Id = s.Id,                     
                    Category = new Category {
                        Id = s.Category.Id,
                        Name = s.Category.Name
                    } 
                });
```
We've got : "Nullable object must have a value" and  we can't test for the moment because we need to rewrite ~ 40 Queries to check the nulls from Left Joins .

The next release will resolve https://github.com/dotnet/aspnetcore/issues/25951 and https://github.com/dotnet/aspnetcore/issues/26174 ?

Because, if you solve these issues and the issues on mobile devices, the Blazor Server Side is almost perfect for Development in big Projects too (my opinion) and we can get over and live with : Development Build time when you change the html, Try Catch Everywhere, UI Clipping and need to use Memory Cache, Login with Redirect and must to use some javascript calls to do it without redirect, Custom Router if you need to have dynamic pages , etc :D

We've implemented a lot of functionalities with enough interactions with javascript  and SQL server queries and seems to be stable. 

There is no much difference in memory or processor resources than in .NET Core 2 (our old platform) . In our case , when there are ~ 20 Users connected  in the same time and they do a lot of interactions  (into an Azure VM D2ds v4 ) the processor is (~5-10 %) and the memory ~ 140 -200 MB. We couldn't test it more for the moment .

**Comment by SteveSandersonMS** [comment]
> The next release will resolve #25951 and #26174 ?

Those two aren't included in the 5.0 release, but I'd be happy to work with you to make sure you have a satisfactory solution. Let's continue the discussion for each of those on their own threads rather than here.

**Comment by SteveSandersonMS** [comment]
Closing as this is resolved in RC2 and the other concerns mentioned above are covered by other issues (which themselves are also resolved).
